internal/fs/dir_handle.go

```diff
@@ -56,36 +55,74 @@ type dirHandle struct {
 	//
 	// GUARDED_BY(Mu)
 	entriesValid bool
+
+	// Condition variable is for signalling whether a fresh set of entries has been fetched.
+	cond *sync.Cond
+
+	// Error during the fetching goroutine must be communicated using this to the main go routine
+	// serving kernel requests.
+	err error
+
+	// Using this as a identification flag to indicate all entries present in the directory
+	// have been fetched already so that the main goroutine does not wait indefinitely for more entries.
+	fetchOver bool
+
+	// To stop the fetching of entries in case of interrupts to main goroutine like ctrl +c
+	// from the user.
+	cancel context.CancelFunc
 }
 
 // Create a directory handle that obtains listings from the supplied inode.
-func newDirHandle(
+func NewDirHandle(
 	in inode.DirInode,
-	implicitDirs bool) (dh *dirHandle) {
+	implicitDirs bool) (dh *DirHandle) {
 	// Set up the basic struct.
-	dh = &dirHandle{
-		in:           in,
+	dh = &DirHandle{
+		In:           in,
 		implicitDirs: implicitDirs,
 	}
 
 	// Set up invariant checking.
 	dh.Mu = locker.New("DH."+in.Name().GcsObjectName(), dh.checkInvariants)
-
+	// Creating a condition variable to indicate events for locking and unlocking dh.Mu mutex.
+	dh.cond = sync.NewCond(dh.Mu)
+	// Using dummy function since nil pointer reference in fs.ReleaseDirHandle if
+	// not populated from fetchEntriesAsync.
+	dh.cancel = dummyCancel
```

Comment:
can we add a nil check in ReleaseDirHandle() instead of assigning a dummyCancel?

Reply:
Added nil check and removed dummycancel.